packages/cohort_sdk/src/cohort.rs

```diff
@@ -144,45 +143,108 @@ impl Cohort {
+                (cpt_snapshot, Vec::<u64>::new())
+            );
+            return (cpt_snapshot, vec![]);
+        }
+
+        let mut cpt_version_min: u64 = u64::MAX;
+        for v in cpt_versions.iter() {
+            if cpt_version_min > *v {
+                cpt_version_min = *v;
+            }
+        }
+        let snapshot_version = std::cmp::max(cpt_snapshot, cpt_version_min);
+        let mut read_vers = Vec::<u64>::new();
+        for v in cpt_versions.iter() {
+            if snapshot_version < *v {
+                read_vers.push(*v);
+            }
+        }
+
+        log::debug!(
+            "select_snapshot_and_readvers({}, {:?}): {:?}",
+            cpt_snapshot,
+            cpt_versions,
+            (snapshot_version, read_vers.clone())
+        );
+        (snapshot_version, read_vers)
+    }
+
+    pub async fn shutdown(&self) {
+        self.agent_services.decision_reader.abort();
+        self.agent_services.state_manager.abort();
+    }
+
+    /// Certifies a candidate in talos and install the statemap if the certification is successful.
+    /// Uses two callbacks
+    ///  - First callback to build the request payload to send to talos for certification
+    ///  - The installer callback will be called to do the out of order install.
+    pub async fn certify<F, Fut, IF, IFut>(
         &self,
-        request: model::CertificationRequest,
-        state_provider: &S,
-        oo_installer: &O,
+        create_certifier_candidate_callback: &F,
+        installer_callback: &IF,
     ) -> Result<model::CertificationResponse, ClientError>
     where
-        S: ItemStateProvider,
-        O: OutOfOrderInstaller,
+        F: Fn() -> Fut,
+        Fut: Future<Output = Result<CertificationCandidateCallbackResponse, String>>,
+        IF: Fn(OOOInstallerPayload) -> IFut,
+        IFut: Future<Output = Result<OutOfOrderInstallOutcome, String>>,
     {
         let span_1 = Instant::now();
-        let response = self.send_to_talos(request, state_provider).await?;
+        // 1. Get the snapshot
+        // 2. Send for certification
+        let response = self.send_to_talos(create_certifier_candidate_callback).await?;
         let span_1_val = span_1.elapsed().as_nanos() as f64 / 1_000_000_f64;
 
         let h_talos = Arc::clone(&self.talos_histogram);
         tokio::spawn(async move {
             h_talos.record(span_1_val * 100.0, &[]);
         });
 
-        if response.decision == Decision::Aborted {
+        if response.decision == Decision::Aborted || response.statemaps.is_none() {
             return Ok(response);
         }
 
-        // system error if we have Commit decision but no safepoint is given
-        let safepoint = response.safepoint.unwrap();
-        let new_version = response.version;
+        let oooinstall_payload = OOOInstallerPayload {
+            xid: response.xid.clone(),
+            version: response.version,
+            safepoint: response.safepoint.unwrap(),
```

Comment:
is unwrap safe here?

Reply:
Yes ideally it should be safe, as we reach here only if we have committed messages with statemaps to install. 

But to make it safer, just in case there is a bug in certifier, I am update the code above as:-

```rust
if response.safepoint.is_none() || response.statemaps.is_none() {
            return Ok(response);
        }
```

Reply:
If there's a bug in the certifier, how can we even safely act on a decision it provided?